openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/storages/builders/ShadowIndexGraphStorageBuilder.java

```diff
@@ -0,0 +1,139 @@
+/*  This file is part of Openrouteservice.
+ *
+ *  Openrouteservice is free software; you can redistribute it and/or modify it under the terms of the 
+ *  GNU Lesser General Public License as published by the Free Software Foundation; either version 2.1 
+ *  of the License, or (at your option) any later version.
+
+ *  This library is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
+ *  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. 
+ *  See the GNU Lesser General Public License for more details.
+
+ *  You should have received a copy of the GNU Lesser General Public License along with this library; 
+ *  if not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+package org.heigit.ors.routing.graphhopper.extensions.storages.builders;
+
+import com.graphhopper.GraphHopper;
+import com.graphhopper.reader.ReaderWay;
+import com.graphhopper.storage.GraphExtension;
+import com.graphhopper.util.EdgeIteratorState;
+import com.graphhopper.util.Helper;
+import org.heigit.ors.routing.graphhopper.extensions.storages.ShadowIndexGraphStorage;
+
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+
+/**
+ * This class builds the shadow index data as a new ShadowGraphStorage.
+ *
+ */
+public class ShadowIndexGraphStorageBuilder extends AbstractGraphStorageBuilder {
+    private ShadowIndexGraphStorage _storage;
+    private Map<Long, Integer> osm_shadowindex_lookup = new HashMap<>();
+    private int max_level = 100;
+    private final int no_data = 30;
+
+    public ShadowIndexGraphStorageBuilder() {
+
+    }
+
+    @Override
+    public GraphExtension init(GraphHopper graphhopper) throws Exception {
+        if (_storage != null)
+            throw new Exception("GraphStorageBuilder has been already initialized.");
+
+        // TODO Check if the shadow index file exists
+        String csvFile = parameters.get("filepath");
+        System.out.print("Shadow Index File: " + csvFile + "\n");
+        readNoiseIndicesFromCSV(csvFile);
+        _storage = new ShadowIndexGraphStorage();
+
+        return _storage;
+    }
+
+    private void readNoiseIndicesFromCSV(String csvFile) throws IOException {
+        BufferedReader csvBuffer = null;
+        try {
+            String row;
+            csvBuffer = new BufferedReader(new FileReader(csvFile));
+            // Jump the header line
+            csvBuffer.readLine();
```

Comment:
Should at least check for null return value when reading from file. Is it even necessary to skip the header line? The test-csv files do not contain a header line, so the first data line is simply ignored.

Reply:
I've adapted it so it's done the same way as in `readNoiseIndicesFromCSV`